datasetinsights/io/source/base.py

```diff
@@ -0,0 +1,42 @@
+import re
+from abc import ABCMeta, abstractmethod
+
+
+class DownloaderRegistry(ABCMeta):
+    registry = {}
+
+    def __new__(cls, name, bases, namespace):
+
+        if "SOURCE_SCHEMA" not in namespace:
+            raise RuntimeError(
+                f"{name} must define a class-level SOURCE_SCHEMA"
+            )
+        new_cls = super().__new__(cls, name, bases, namespace)
+        if namespace["SOURCE_SCHEMA"] != "":
+            DownloaderRegistry.registry[namespace["SOURCE_SCHEMA"]] = new_cls
+        return new_cls
+
+    @classmethod
+    def find(cls, source_uri):
```

Comment:
How do we enable automatically register new protocols if a new downloader (such as `s3://`), is added to the repo?

Reply:
We would need to refactor 2 different places this class we would need to add the s3 to the pattern of the re.compile method and in the SourceURI class that validate the source-uri.
The downloader is automatically registered if it is a subclass of DatasetDownloader if that is what your question is?

Reply:
Since we already know the protocol (gs://, http://...) of all subclasses, can we dynamically determine from source URI to each downloader class without making code changes?

Some suggestions:

```
protocols = [d.PRORO for d in DatasetDownloader.subclasses()]
pattern = re.compile("|".join( protocols))
# or 
protocol_to_cls = {d.PRORO: d for d in DatasetDownloader.subclasses()}
protocol = parse_source_uri_to_get_protocol(...)
downloader_cls = protocol_to_cls[protocol]
```